### ISSUE TITLE
Flaky behavior of GT axe durability when harvesting brittle charcoal

### ISSUE THREAD

**Issue by MauveCloud** [Blood-Asp/GT5-Unofficial#439]
StragaSevera mentioned it with regard to flint axes in the GT5u thread, but I've also seen a similar thing happen with a GT iron axe - the durability bar shows it goes down, then comes back up.  Once the durability is below about 2600 (not certain of exact threshold), the axe breaks for about half a second, but then reappears with durability only down by 25.

**Comment by Timeslice42** [comment]
Sounds like a desync between server and client.

**Comment by MauveCloud** [comment]
Two problems with that theory:
1. I was getting it in single-player (I know it's technically running as a server on the same machine as the client, but that should make it highly resistant to such desync issues)
2. It happened repeatedly as I went through the brittle charcoal - with a diamond GT axe, it started temporarily breaking at about durability 5800 (and with InvTweaks "sort on pickup" enabled I often had to open my inventory to transfer it back to the hotbar)

**Comment by Timeslice42** [comment]
In modded minecraft client/server desyncs rarely have anything to with network traffic (or multiplayer). The client displays on screen what it thinks happened based on your actions. At some point in the (near) future the server responds with what actually happened and the client updates accordingly. Normally this is all without the player's knowledge, but on occasion what the client thinks happened is not what the server says happened, and the server is always authoritative.

There are even a few ways you can see this in vanilla minecraft SSP. When trying to place a block on the cieling at your maximum reach you will briefly see it get placed, and then it will disappear. This is because the client and the server aren't using the same criteria for block placement (I'm hoping this gets fixed in 1.8+, but it's been a long-standing bug). You can also see this with any tool/weapon that has unbreaking on it, especially if you're using the armor status HUD mod which shows the numerical durability of an item you're using. This is because the server and the client don't agree on when unbreaking prevents damage (as they aren't rolling the same pair of dice, to use a metaphor).

The fact that inventory tweaks is a client side mod is why you'll see it move the tool even when it doesn't actually break, as it sees what you see.

Really the only part of this that @Blood-Asp should probably take a look at is why in the world the client thinks the axe is taking 2000+ durability damage. That, and the fact that the shovel is supposed to be the operational tool for the charcoal pit...

**Comment by MauveCloud** [comment]
> Really the only part of this that @Blood-Asp should probably take a look at is why in the world the client thinks the axe is taking 2000+ durability damage.

That might suggest that it's getting partially confused with one of the other reinforced stone blocks - a Bronzeplate Reinforced Block, which has the same block id as brittle charcoal but metadata 0 instead of 4, would take 3000 durability to mine with a GT metatool. (though why Blood Asp decided to implement brittle charcoal with the same class and block id as reinforced blocks is beyond me)

> That, and the fact that the shovel is supposed to be the operational tool for the charcoal pit...

Actually, since brittle charcoal is converted from logs, using an axe seems logical to me.

**Comment by Timeslice42** [comment]
Every charcoal pit I've ever seen (in minecraft, other games, and real life) has used a shovel to dig up the charcoal. Mainly because it's pretty hard to dig at all with an axe.

**Comment by Blood-Asp** [comment]
The coded tool is an axe: https://github.com/Blood-Asp/GT5-Unofficial/blob/experimental/src/main/java/gregtech/common/blocks/GT_Block_Reinforced.java#L67 just somehow does not work...

I made it part of the reinforced blocks, since i had free blocks there. Otherwise i would have to reserve new block IDs from MC what should not be done.

My theory is, that the client only checks the default dura of an block and not dependend on the meta. So the issue might be on side of the GT meta tools and not in the block. Maybe, i did not look into it so far.

**Comment by MauveCloud** [comment]
> My theory is, that the client only checks the default dura of an block and not dependend on the meta. So the issue might be on side of the GT meta tools and not in the block. Maybe, i did not look into it so far.

I see a small problem with that theory: the meta tool code doesn't check the metadata of the block, the getBlockHardness function (in GT_Block_Reinforced.java) does.

**Comment by axlegear** [comment]
Can confirm this happens with steel pickaxes or flint pickaxes on brittle charcoal on our HAM server.  It's quite wierd to break the same pickaxe fifty times.  XD

**Comment by MauveCloud** [comment]
Another weirdness is that different-material axes apparently start breaking and unbreaking at different durabilities.  Based on my previous comments, an iron axe does that around 2600, a diamond one does it at about 5800, and I just tried it with a flint axe, which starts doing it right at 3000.

Edit: correction: due to the higher speed of the diamond axe, it often mines two or more brittle charcoal in quick succession instead of just one, hence why it seemed to start breaking and unbreaking around durability 5800, but that didn't happen when mining a single brittle charcoal until durability 3000, which I confirmed is the same for iron and certus quartz.

**Comment by MauveCloud** [comment]
I tried temporarily adding logging to GT_Block_Reinforced.getBlockHardness to show what metadata it was finding, and all the lines showed it was getting metadata 4.  Some additional weird behavior: an IC2 mining drill mines brittle charcoal faster than an IC2 chainsaw, and (at least in survival) a GT flint pickaxe also works to mine brittle charcoal, even though the block is coded to make an axe the correct tool to mine it.

Edit: Correction: I tried again, and found a line where it found metadata 0 (and returned 60.0 for hardness).  Investigating further to try to determine why.

Edit 2: the function gets called on an air block, which has metadata 0, before being called again at least once with the appropriate block there.
